pkg/image/admission/admission.go

```diff
@@ -0,0 +1,316 @@
+	}
+}
+
+// Stop stops the reflectors used by the plugin.
+func (a *imageLimitRangerPlugin) Stop() {
+	if a.stopChan != nil {
+		close(a.stopChan)
+		a.stopChan = nil
+	}
+}
+
+// Admit invokes the admission logic for checking against LimitRanges.
+func (a *imageLimitRangerPlugin) Admit(attr kadmission.Attributes) error {
+	if !a.supports(attr.GetResource()) || attr.GetSubresource() != "" {
+		return nil // not applicable
+	}
+
+	// If the namespace is not set on the object ensure it gets set.
+	if om, err := kapi.ObjectMetaFor(attr.GetObject()); err == nil {
+		if om.Namespace == "" {
+			glog.V(5).Infof("%s: namespace was not set for object, defaulting to %s", pluginName, attr.GetNamespace())
+			om.Namespace = attr.GetNamespace()
+		}
+	}
+
+	return a.limitRanger.Admit(attr)
+}
+
+// supports is a helper to that returns true if the resource is supported by the plugin.
+func (a *imageLimitRangerPlugin) supports(resource unversioned.GroupResource) bool {
+	return resource == imageapi.Resource("imagestreammappings") ||
+		resource == imageapi.Resource("imagestreamtags") ||
+		resource == imageapi.Resource("imagestreams")
+}
+
+// SetOpenshiftClient provides init functionality for anything that requires the OpenShift client.
+func (a *imageLimitRangerPlugin) SetOpenshiftClient(c client.Interface) {
+	a.osClient = c
+
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return a.osClient.ImageStreams(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return a.osClient.ImageStreams(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&imageapi.ImageStream{},
+		a.imageStreamStore,
+		0,
+	)
+	a.imageStreamReflector = reflector
+
+	a.Run()
+}
+
+func (a *imageLimitRangerPlugin) Validate() error {
+	if a.osClient == nil {
+		return fmt.Errorf("%s needs an Openshift client", pluginName)
+	}
+	return nil
+}
+
+// imageLimitRanger is the limit range implementation that checks resource against the
+// image limit ranges.
+func (a *imageLimitRangerPlugin) imageLimitRanger(limitRange *kapi.LimitRange, kind string, obj runtime.Object) error {
+	if !limitAppliesToImages(limitRange) {
+		return nil
+	}
+
+	var images map[string]imageapi.Image
+
+	switch obj.(type) {
+	case *imageapi.ImageStreamMapping:
+		ism, ok := obj.(*imageapi.ImageStreamMapping)
+		if !ok {
+			glog.V(4).Infof("%s: unable to convert obj to ImageStreamMapping: %#v", pluginName, obj)
+			return nil
+		}
+		images = map[string]imageapi.Image{ism.Image.Name: ism.Image}
+	case *imageapi.ImageStreamTag:
+		tag, ok := obj.(*imageapi.ImageStreamTag)
+		if !ok {
+			glog.V(4).Infof("%s: unable to convert obj to ImageStreamTag: %#v", pluginName, obj)
+			return nil
+		}
+
+		if tag.Tag == nil || tag.Tag.From == nil {
+			glog.V(4).Infof("%s: ignoring ist with no tag reference to follow: %v", pluginName, tag)
+			return nil
+		}
+
+		if tag.Tag.Reference {
```

Comment:
Why shall we ignore references? They allow to pull source istag as well.

If justified, it deserves at least an explanation and another change to quota counting.